tools/distribution/verify.py

```diff
@@ -0,0 +1,203 @@
+        """The test config class"""
+        return distrotest.DistroTestConfig
+
+    @property
+    def testfile(self) -> pathlib.Path:
+        """Path to a testfile to run inside the test containers"""
+        return pathlib.Path(self.args.testfile)
+
+    @cached_property
+    def tests(self) -> dict:
+        """A dictionary of tests and test configuration, filtered according
+        to provided args
+        """
+        _ret = {}
+        for name, config in self.config.items():
+            if self.filter_distributions and name not in self.filter_distributions:
+                continue
+            _ret[name] = self.get_test_config(config["image"])
+            _ret[name].update(config)
+            _ret[name]["packages"] = self.get_test_packages(_ret[name]["type"], _ret[name]["ext"])
+        return _ret
+
+    def add_arguments(self, parser: argparse.ArgumentParser) -> None:
+        super().add_arguments(parser)
+        parser.add_argument(
+            "testfile",
+            help="Path to the test file that will be run inside the distribution containers")
+        parser.add_argument(
+            "config", help="Path to a YAML configuration with distributions for testing")
+        parser.add_argument("packages", help="Path to a tarball containing packages to test")
+        parser.add_argument(
+            "--keyfile",
+            "-k",
+            help="Specify the path to a file containing a gpg key for verifying packages.")
+        parser.add_argument(
+            "--distribution",
+            "-d",
+            nargs="?",
+            help="Specify distribution to test. Can be specified multiple times.")
+        parser.add_argument(
+            "--rebuild", action="store_true", help="Rebuild test images before running the tests.")
+
+    async def check_distros(self) -> None:
+        """Check runner"""
+        for name, config in self.tests.items():
+            self.log.info(
+                f"[{name}] Testing with: "
+                f"{','.join(p.name for p in config['packages'])}")
+            for i, package in enumerate(config["packages"]):
+                await self.run_test(name, config["image"], package, (i == 0 and self.rebuild))
+
+    def get_test_config(self, image: str) -> dict:
+        """Get the type/ext config for a given image name"""
+        return self.test_config.get_config(image)
+
+    def get_test_packages(self, type: str, ext: str) -> list:
+        """Get the packages to test for a given type/ext"""
+        return self.test_config.get_packages(type, ext)
+
+    async def on_checks_complete(self) -> int:
+        """Cleanup and return the test result"""
+        await self._cleanup_test()
+        await self._cleanup_docker()
+        return await super().on_checks_complete()
+
+    async def run_test(self, name: str, image: str, package: pathlib.Path, rebuild: bool) -> None:
+        """Runs a test for each of the packages against a particular distro"""
+        if self.exiting:
```

Comment:
The need to manual check for exiting here feels fragile.

Reply:
in the scheme of things this is defensive, and not strictly required

`self.exiting` is `True` in response to `KeyboardExit` - this triggers a cleanup, and this opens a very brief window where other co processes can still be running - this basically prevents it starting a new test during that window

i think we can ~safely remove, but in the rare occasion that this is hit, it would just make it keyboard exiting slower

i think i would rather keep, but ambivalent enough to change it

Reply:
i have left for now, lmk if that is a blocker

Reply:
Is there no way to add a co-routine start hook to do this more generically? I'm just a little concerned that we have to do this by hand. The reason why it is done makes sense.

Reply:
its worth mentioning that these tests are run serially, so at least with the current implementation, its mostly just the cleanup that needs to be worried about on exit

Reply:
as a general point working `asyncio` its better to have a routine/loop that knows how to exit gracefully when it should, than to have routine that forces cancellation/closure on exit

a good example of this is the `submit_task` in the `parallel` pr

in my initial implementation in its cancel method i had

```python
self.submit_task.cancel()
try:
    await self.submit_task
except asyncio.CancellationError:
    # not always hit - deps on timing of coroutines created by submit_task
    pass
```

Nothing i could do could seemingly prevent it from raising the cancellation error

Then i added a `closing` lock and told the submit_task to bail gracefully when locked and replaced above code with simply

```
await self.submit_task
```

and everything was happy

Reply:
one thing im wondering from this discussion - whether in the case of async checkers the `exiting` prop should use an `asyncio.Lock` 

the asyncio locks are not thread-safe (and dont need to be) so its not immediately clear what the advantages of using the asyncio sync primitives is

from my testing it seems like they offer a better guarantee of synchronization in terms of coroutine execution - so my ~guess is that they are run with higher priority than normal tasks

they can also be used synchronously so we could theoretially just use it for all checkers albeit slightly differently altho im a little uncomfortable with using anything asyncio in non-async code

Reply:
https://docs.python.org/3/library/asyncio-sync.html

Reply:
actually `Lock` cant be used synchronously, just queue